qiskit_nature/problems/second_quantization/lattice/lattice/hyper_cubic_lattice.py

```diff
@@ -0,0 +1,214 @@
+            else:
+                raise TypeError(
+                    "size mismatch, "
+                    f"`edge_parameter`: {len(edge_parameter)}, `size`: {self.dim}."
+                    "The length of `edge_parameter` must be the same as that of size."
+                )
+
+        self.edge_parameter = edge_parameter
+        # onsite parameter
+        self.onsite_parameter = onsite_parameter
+
+        # boundary condition
+        if isinstance(boundary_condition, str):
+            boundary_condition = (boundary_condition,) * self.dim
+        elif isinstance(boundary_condition, tuple):
+            if len(boundary_condition) != self.dim:
+                raise ValueError(
+                    f"The length of `boundary_condition` must be the same as that of size, {self.dim}."
+                )
+
+        self.boundary_conditions = boundary_condition
+
+        graph = PyGraph(multigraph=False)
+        graph.add_nodes_from(range(np.prod(size)))
+
+        # add edges excluding the boundary edges
+        coordinates = list(product(*map(range, size)))
+        base = np.array([np.prod(size[:i]) for i in range(self.dim)], dtype=int)
+        for coord in coordinates:
+            for i in range(self.dim):
+                if coord[i] != size[i] - 1:
+                    node_a = np.dot(coord, base)
+                    node_b = node_a + base[i]
+                    graph.add_edge(node_a, node_b, edge_parameter[i])
+
+        # add self-loops
+        for node_a in range(np.prod(size)):
+            graph.add_edge(node_a, node_a, onsite_parameter)
+
+        # depend on the boundary condition
+        self.boundary_edges = []
+        for i in range(self.dim):
+            # add edges when the boundary condition is periodic.
+            # when the boundary condition in the i-th direction is periodic,
+            # it makes sense only when size[i] is greater than 2.
+            if boundary_condition[i] == "periodic":
+                if size[i] <= 2:
+                    continue
+                size_list = list(size)
+                size_list[i] = 1
+                coordinates = list(product(*map(range, size_list)))
+                for coord in coordinates:
+                    node_b = np.dot(coord, base)
+                    node_a = node_b + base[i] * (size[i] - 1)
+                    if node_a < node_b:
+                        graph.add_edge(node_a, node_b, edge_parameter[i])
+                        self.boundary_edges.append((node_a, node_b))
+                    elif node_a > node_b:
+                        graph.add_edge(node_b, node_a, edge_parameter[i].conjugate())
+                        self.boundary_edges.append((node_a, node_b))
+            elif boundary_condition[i] != "open":
+                raise ValueError(
+                    f"Invalid `boundary condition` {boundary_condition[i]} is given."
+                    "`boundary condition` must be `open` or `periodic`."
+                )
```

Comment:
Thank you for the comment. When we make a class that generates more general translational invariant lattices which have complicated edges, such an extraction will be needed. However, now since `HyperCubicLattice` class is one of the most general ones to generate translational invariant lattices, it seems that we don't need to extract the process here right now.

Reply:
Instead, I added comments to improve readability.

Reply:
I reviewed the code and have wrapped some code blocks as functions.